File: rogue-server/tracker/trackers/action_tracker.py
from abc import ABC

from tracker.tracker_template import Tracker


class ActionTracker(Tracker, ABC):

    def __init__(self, data):
        super().__init__(data)

        self.last_attack = 0

        self.sneaking = False
        self.digging = False

    def handle(self, event):
        if event['type'] == 'in_use_entity':
            self.last_attack = self.data.ticks_existed
        elif event['type'] == 'in_entity_action':
            action = event['packet']['playerAction']

            if action == 'START_SNEAKING':
                self.sneaking = True
            elif action == 'STOP_SNEAKING':
                self.sneaking = False
        elif event['type'] == 'in_dig':
            action = event['packet']['digType']

            if action == 'START_DESTROY_BLOCK':
                self.digging = True
            elif action == 'STOP_DESTROY_BLOCK' or action == 'ABORT_DESTROY_BLOCK':
                self.digging = False

    def is_attacking(self):
        return self.data.ticks_existed - self.last_attack < 1

    def is_digging(self):
        return self.digging

File: rogue-server/src/main/java/org/hostile/anticheat/tracker/impl/player/ActionTracker.java
package org.hostile.anticheat.tracker.impl.player;

import lombok.Getter;
import org.hostile.anticheat.data.PlayerData;
import org.hostile.anticheat.event.PacketEvent;
import org.hostile.anticheat.packet.inbound.WrappedPacketPlayInBlockDig;
import org.hostile.anticheat.packet.inbound.WrappedPacketPlayInEntityAction;
import org.hostile.anticheat.packet.inbound.WrappedPacketPlayInUseEntity;
import org.hostile.anticheat.tracker.Tracker;

@Getter
public class ActionTracker extends Tracker {

    private int lastAttack;

    private boolean sneaking;
    private boolean digging;

    public ActionTracker(PlayerData data) {
        super(data);
    }

    @Override
    public void handle(PacketEvent event) {
        if (event.getPacket() instanceof WrappedPacketPlayInUseEntity) {
            WrappedPacketPlayInUseEntity packet = (WrappedPacketPlayInUseEntity) event.getPacket();

            if (packet.getAction() != WrappedPacketPlayInUseEntity.EntityUseAction.ATTACK) {
                return;
            }

            lastAttack = data.getTicksExisted() + 1;
        } else if (event.getPacket() instanceof WrappedPacketPlayInEntityAction) {
            WrappedPacketPlayInEntityAction packet = (WrappedPacketPlayInEntityAction) event.getPacket();

            switch (packet.getAction()) {
                case START_SNEAKING:
                    sneaking = true;
                    break;
                case STOP_SNEAKING:
                    sneaking = false;
                    break;
            }
        } else if (event.getPacket() instanceof WrappedPacketPlayInBlockDig) {
            WrappedPacketPlayInBlockDig packet = (WrappedPacketPlayInBlockDig) event.getPacket();

            switch (packet.getPlayerDigType()) {
                case START_DESTROY_BLOCK:
                    digging = true;
                    break;
                case STOP_DESTROY_BLOCK:
                case ABORT_DESTROY_BLOCK:
                    digging = false;
                    break;
            }
        }
    }

    public boolean isAttacking() {
        return data.getTicksExisted() == lastAttack;
    }

    public boolean hasAttackedRecently() {
        return Math.abs(data.getTicksExisted() - lastAttack) <= 3;
    }
}
